### PR TITLE
Refactor/better aggregate meters

### PR DESCRIPTION
**Pull request by macintoshpie**
#### Any background context you want to provide?
NREL's ESTCP project includes sample data with spotty meter data which doesn't work well with BETTER by default. There were a few issues with the data that presented new challenges communicating with BETTER:
  1. There were _a lot_ of meter readings and the BSync file created ended up being much too large (see the referenced issue) because SEED is currently dumping meters as they are into the file. We needed to reduce the readings included in the BSync file
  2. The data was pretty incomplete -- there were some missing months and BETTER requires at least 12 _continuous_ months of readings (it assumes usage of 0 for missing months which wrecks the models)

#### What's this PR do?
Adds an option to the BETTER analysis to "preprocess" meters. This should probably be just a temporary implementation because not much thought/knowledge about meters was put into it.
This "preprocessing" basically:
- bins meter readings into monthly readings (i.e. aggregate readings or disperse them)
  - e.g. if you have multiple 15-minute-interval readings it will sum them up for the month
  - e.g. if a reading _spans_ or _straddles_ a month, the algorithm splits the energy usage according to the fraction of the time the reading covers a particular month. This is a big ⚠️ that this is not a great/perfect solution; it assumes even distribution of usage throughout the meter reading interval which is likely not true (especially for readings that span a seasonal change)
- drops outliers (was experiencing some weird months due to binning - especially on the starting/ending boundaries)
- interpolates missing months

#### How should this be manually tested?
Run better analysis with and without preprocessing. It should work.

#### What are the relevant tickets?
#2917 

#### Screenshots (if appropriate)

New option on BETTER analysis
Screen Shot 2021-10-01 at 3 07 25 PM

Description of new option
Screen Shot 2021-10-01 at 3 07 34 PM

### CODE REVIEW

File: seed/analysis_pipelines/better/client.py [hunk 1]
```diff
@@ -19,6 +19,24 @@ class BETTERClient:
     def __init__(self, token):
         self._token = f'Token {token}'
 
+    def test_token(self):
```
Comment (macintoshpie) [inline seed/analysis_pipelines/better/client.py:22]: This change is unrelated to the meter stuff -- we were using the /buildings/ api to check if tokens were valid when creating the analysis but as we created buildings this endpoint got much too slow. We should ask BETTER team to create a token endpoint for testing. For now we just use their weathers API and check if we got a 401

File: seed/analysis_pipelines/better/pipeline.py [hunk 2]
```diff
@@ -93,9 +100,9 @@ def _prepare_analysis(self, property_view_ids, start_analysis=False):
 
         # ping BETTER to verify the token is valid
         client = BETTERClient(organization.better_analysis_api_key)
-        _, errors = client.get_buildings()
-        if errors:
-            message = '; '.join(errors)
+        success = client.test_token()
```
Comment (macintoshpie) [inline seed/analysis_pipelines/better/pipeline.py:103]: 🚤

File: seed/analysis_pipelines/better/pipeline.py [hunk 3]
```diff
@@ -139,6 +146,54 @@ def _start_analysis(self):
...
+            )
+        )
+        for meter in meters:
+            # filtering on readings >= 1.0 b/c BETTER flails when readings are less than 1 currently
+            readings = meter.meter_readings.filter(reading__gte=1.0).order_by('start_time')
+            if readings.count() >= 12:
+                selected_meters_and_readings.append({
+                    'meter_type': meter.type,
+                    'readings': readings,
+                })
```
Comment (macintoshpie) [inline seed/analysis_pipelines/better/pipeline.py:192]: this matches the original implementation of gathering meters and readings from before

File: seed/analysis_pipelines/better/pipeline.py [hunk 4]
```diff
@@ -139,6 +146,54 @@ def _start_analysis(self):
...
+            monthly_readings = aggregate_meter_readings(meter_readings)
+            monthly_readings = reject_outliers(monthly_readings)
+            # filtering on readings >= 1.0 b/c BETTER flails when readings are less than 1 currently
+            monthly_readings = [reading for reading in monthly_readings if reading.reading >= 1.0]
+            monthly_readings = interpolate_monthly_readings(monthly_readings)
+            if len(monthly_readings) >= 12:
+                selected_meters_and_readings.append({
+                    'meter_type': meter_type,
+                    'readings': monthly_readings
+                })
```
Comment (macintoshpie) [inline seed/analysis_pipelines/better/pipeline.py:174]: Section for preprocessing meters. There are some interesting differences from our original method:
1. we combine readings from the same meter type (instead of keeping different meters separate)
2. we "bin" meters into calendar months
3. we reject outliers
4. we interpolate missing months

**Review by Ryoken — APPROVED** [review]
My only comment would be to include the translation for the HTML tooltip.  Looks good otherwise!

**Comment by macintoshpie** [comment]
I will make a ticket to translate the tooltips!
